### ISSUE TITLE
nlog for mono 2

### ISSUE THREAD

**Issue by ujaroe** [NLog/NLog#1644]
Hi,
what's the last version of Nlog for Mono 2 and where can I get it?

I'd like to use Nlog with Unity3D, which unfortunately uses a rather outdated version of Mono.

I tried version 4.3.8 for .Net 3.5, which first appears to work. However, after writing to a file Nlog stops with a NotImplementedException which is raised from class System.Runtime.InteropServices/Marshal in method GetHRForLastWin32Error.

I tried the old Mono 2 version from codeplex, but it seems to fail to find the config file. As far as I can tell, I need to use the NLog.dll.nlog alternative. This may not exist in this old version.

Thanks for any advice!

**Comment by 304NotModified** [comment]
The mono 2 version is currently as source only. You need to compile the mono.sln

**Comment by ujaroe** [comment]
Okay, thanks!

But - how would I compile mono.sln, for Mono 2? I see the defines and output paths, but Nlog.mono, when opened, targets .Net Framework 4.5. Should I change it to .Net 3.5?

**Comment by 304NotModified** [comment]
> But - how would I compile mono.sln, for Mono 2? I see the defines and output paths, but Nlog.mono, when opened, targets .Net Framework 4.5. Should I change it to .Net 3.5?

good point. I think you should. Let me know if can't get it compiling. 

FYI the current mono solution is build and unit tested every commit/change on Travis

**Comment by ujaroe** [comment]
Unfortunately build of nlog.mono fails in RegistryLayoutRenderer.cs, line 93 - "RegistryView" is not known.

**Comment by 304NotModified** [comment]
That's strange. Will check that. Do you know if mono 4 = net45?

**Comment by ujaroe** [comment]
Yes, as far as I can tell. See http://www.mono-project.com/docs/about-mono/releases/4.0.0/

**Comment by 304NotModified** [comment]
Sorry about the late reply. I guess this is still an issue?

**Comment by ujaroe** [comment]
Yes, unfortunately. Can I be of help?

**Comment by 304NotModified** [comment]
I will try something, too bad Mono is currently broken on my machine.

Could you google what the problem is with the AuthorizationRule on Mono?

**Comment by 304NotModified** [comment]
Any experience with Mono running on Windows?

Installed mono, but even after reboot getting the error:

```

Error	CS0246	The type or namespace name 'Mono' could not be found (are you missing a using directive or an assembly reference?)	NLog.mono	X:\nlog\NLog\src\NLog\Internal\FileAppenders\UnixMultiProcessFileAppender.cs	51	Build	Active

```

**Comment by 304NotModified** [comment]
it Mono.Posix. Cannot find the mono2 compatible one

**Comment by 304NotModified** [comment]
See https://github.com/NLog/NLog/pull/1682

**Comment by ujaroe** [comment]
Don't know if it helps to fix the problem - but I had a look at the Mono sources which are used by Unity3D (here https://github.com/Unity-Technologies/mono/blob/unity-staging/mcs/class/corlib/System.Security.AccessControl)

Apparently the exception is thrown, because accessMask == 0 in the constructor of class AuthorizationRule. which is called as the base class constructor by class AccessRule, which in turn is called as base class from class MutexAccessRule, where accessMask is unfortunately hard coded as 0. MutexAccessRule itself seems to be called from NLog.Internal.FileAppenders.BaseFileAppender.CreateSharableMutex. Can this be changed at all?

**Comment by 304NotModified** [comment]
> Can this be changed at all?

I think we could change it, at least for Mono :)

**Comment by ujaroe** [comment]
Anything new on this? Anything I could do?

**Comment by 304NotModified** [comment]
Well I didn't found  Mono.Posix for .NET 3.5 and therefor I'm stuck.

**Comment by snakefoot** [comment]
Would a Mono.Posix ver. 2 work ?

[Mono.Posix.zip](https://github.com/NLog/NLog/files/545676/Mono.Posix.zip)

Taken from here:

http://www.go-mono.com/mono-downloads/

**Comment by ujaroe** [comment]
You could download and install an older Mono from http://download.mono-project.com/archive.

I don't know if there will be a problem later on with Unity3D, but there is a mono.posix.dll for CLR 2.0/3.5 inside the latest v3 (3.12.1). You can find it in C:\Program Files (x86)\Mono\lib\mono\2.0 after installation.

**Comment by 304NotModified** [comment]
Will look at this now. I guess the version of @snakefoot will work, Thanks!

**Comment by 304NotModified** [comment]
Almost there :)

**Comment by 304NotModified** [comment]
> if [ "$MONO_V" = "MONO2" ]; then
  nuget restore src/NLog.mono2.sln
  nuget install xunit.runners -Version 1.9.2 -OutputDirectory testrunner
  xbuild /p:Configuration=Release src/NLog.mono2.sln
  mono ./testrunner/xunit.runners.1.9.2/tools/xunit.console.exe      "./build/bin/Release/Mono 2.x/NLog.UnitTests.dll"
fi
./build-mono.sh: line 5: nuget: command not found

Someone any idea how to fix? See https://github.com/NLog/NLog/pull/1682 and https://travis-ci.org/NLog/NLog/jobs/170255542

**Comment by snakefoot** [comment]
Is nuget available?

https://github.com/ekonbenefits/dynamitey/tree/master/.nuget

Was mentioned here:

http://stackoverflow.com/questions/16751772/how-do-i-use-travis-ci-with-c-sharp-or-f

**Comment by 304NotModified** [comment]
> Does NuGet support Mono?
The command-line tool, nuget.exe, builds and runs under Mono 3.2+ and can create packages in Mono.

https://docs.nuget.org/ndocs/policies/nuget-faq

No mono 2 support. But will try, dunno why it won't work currently.

**Comment by snakefoot** [comment]
Seems that you just need the Microsoft.Build.dll (See my first link in previous comment), and then you can run NuGet on Mono2

http://stackoverflow.com/questions/11301613/how-to-run-nuget-on-mono-2-10-9

**Comment by 304NotModified** [comment]
Its almost working

**Comment by 304NotModified** [comment]
I think

**Comment by 304NotModified** [comment]
got now: 

>  The contextual keyword `var' may only appear within a local variable declaration

**Comment by 304NotModified** [comment]
unit tests,

1497 total, 410 failed, 13 skipped, took 105.738 seconds

don't get it. See https://travis-ci.org/NLog/NLog/jobs/170899261

**Comment by 304NotModified** [comment]
anyone time to look at this?

**Comment by snakefoot** [comment]
Seems Mono ver. 2 doesn't support this call:

NLog.Internal.FileAppenders.BaseFileAppender.CreateSharableMutex

MutexAccessRule throws the following exception: 

System.ArgumentOutOfRangeException : Argument is out of range.

Which is caused by Mono ver. 2 not having a fully working MutexAccessRule.

The problem seems to have been fixed with this commit to Mono (Correctly handles MutexRights eventRights)

https://github.com/mono/mono/commit/f669650fdbd4c97895db9ac4fb64263a47bc2d8d

**Comment by 304NotModified** [comment]
> The problem seems to have been fixed with this commit to Mono (Correctly handles MutexRights eventRights)

But that isn't mono 2 right?

**Comment by snakefoot** [comment]
Know very little about Mono's eagerness to maintain old release-branches. Just like to find stuff on Google :).

But when looking at the commit then it is certainly not targeted for Mono ver. 2, and since the build fails with the mentioned exception, it has probably not yet reached Mono ver. 2.

**Comment by 304NotModified** [comment]
yes that seems logical. Thx!

**Comment by 304NotModified** [comment]
Sorry, this is not trivial to fix. Won't expect that it will run with mono 2 soon.

**Comment by 304NotModified** [comment]
From a tweet I understand NLog 4.3.7 was working correctly? 

https://twitter.com/liam_cary/status/804263394245165058

**Comment by 304NotModified** [comment]
changes in NLog 4.3.8: https://github.com/NLog/NLog/milestone/28?closed=1

**Comment by snakefoot** [comment]
Maybe the use of HashSet in FilePathLayout. It is kind of new container. Only available from .NET 3.5 (MONO 3)

**Comment by snakefoot** [comment]
The funny part is that the tweet points back to this thread, where the topic starter tells that 4.3.8 is not working :)

**Comment by 304NotModified** [comment]
> Though no idea what version he was upgrading from.

Yes indeed. I think the problem is in the mutex (which isn't added in NLog 4.3.8?)

**Comment by 304NotModified** [comment]
The archive mutex has been added in NLog 4.3.9 (https://github.com/NLog/NLog/pull/1646)

**Comment by snakefoot** [comment]
@ujaroe NLog ver. 4.4.1 should now perform runtime check whether named mutex is available, and should not throw ArgumentOutOfRangeException.

**Comment by 304NotModified** [comment]
We need to merge this with https://github.com/NLog/NLog/pull/1682

**Comment by ujaroe** [comment]
The problems with 4.3.8 may be related to the actual Unity3D runtime I was using.

Unfortunately I was'nt able to take part in the discussion related to Mono2 itself. However, I can confirm now, that NLog 4.4.1 from Nuget for .Net 3.5 works (Unity3D 5.5.0)!

Thank you so much!

**Comment by ujaroe** [comment]
Oh - no. Have to correct myself. The out-of-box version gives this exception:
System.NotImplementedException: The requested feature is not implemented. (Int32  GetHRForLastWin32Error())
However, it seems I can avoid this by compiling myself according to [https://github.com/NLog/NLog/issues/1644#issuecomment-248829571](url)

**Comment by snakefoot** [comment]
Try and enable ForceManaged=true on the FileTarget.

**Comment by ujaroe** [comment]
With >ForcedManaged="true"< on the file target, the out-of-box-version 4.4.1 (.Net 3.5) appears to work correctly so far.

Thanks again!

**Comment by snakefoot** [comment]
Created #1893 to always use ForcedManaged on MONO.

**Comment by snakefoot** [comment]
@304NotModified Guess this has been resolved with NLog 4.4.2, as the mono-runtime-checks makes the NET3_5 nuget-package workable for MONO2.

**Comment by 304NotModified** [comment]
Yes indeed, but I also trying to get mono2 running on travis, almost there I think

https://github.com/NLog/NLog/pull/1682
